### PR TITLE
Branch and bound

### PR DESCRIPTION
**Pull request by threelittle87**
Please include a summary of the change.
Please also include relevant motivation and context (if applicable).
Put 'x' in between square brackets to mark an item as complete.

- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Comments in areas I changed are up to date
- [x] I have added comments to hard-to-understand areas of my code
- [x] I have made corresponding changes to the README.md
- [x] I have made corresponding changes to the DIRECTORY.md

### CODE REVIEW

**Comment by threelittle87** [comment]
The non-tested code detected by codecov are the constructors for the custom exception, only one of all the constructors is used in the algorithm implementation, the others are added just to fix the "update this implementation of 'ISerializable' to conform to the recommended serialization pattern." codacy issue. If the code passes the codacy review, it will fail the codecov review and vice versa. Any suggestions?

**Comment by threelittle87** [comment]
The codacy review is saying that my custom exception class is not following the recommended exception class implementation. I noticed that the ItemNotFoundException class in Utilities\Exceptions is not following the recommendation as well. Therefore I assume it is fine to ignore this codacy issue.

File: Algorithms/Knapsack/BranchAndBoundNode.cs
Comment (siriak) [inline Algorithms/Knapsack/BranchAndBoundNode.cs:86]: How is it used? And this method should return non-nullable BranchAndBoundNode since it never returns null actually

File: Algorithms/Knapsack/BranchAndBoundNode.cs
Comment (siriak) [inline Algorithms/Knapsack/BranchAndBoundNode.cs:64]: Was this ported from Java? Please use auto properties https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/auto-implemented-properties

File: Utilities/Exceptions/BranchAndBoundNullParentException.cs
Comment (siriak) [inline Utilities/Exceptions/BranchAndBoundNullParentException.cs:8]: I don't see when it should be used (semantically). Probably it should just be removed.

File: Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs
Comment (siriak) [inline Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs:185]: Could you share sources for that? Looks like a heuristic to me.

File: Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs
Comment (siriak) [inline Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs:130]: This line doesn't make sense to me, please review working with nulls and return type of GetParent()

File: Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs
Comment (siriak) [inline Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs:65]: Please reduce nesting
if (parent.GetLevel() > items.Length)
{
continue;
}

File: Algorithms.Tests/Knapsack/BranchAndBoundKnapsackSolverTests.cs
Comment (siriak) [inline Algorithms.Tests/Knapsack/BranchAndBoundKnapsackSolverTests.cs:30]: Please use FluentAssertions

**Review by siriak — COMMENTED** [review]
Thanks for contributing! Very good overall and only need to polish a bit to make idiomatic C# :D

File: Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs
Comment (threelittle87) [reply Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs:185]: Sorry my English is bad. From what I get from wikipedia "A heuristic algorithm is one that is designed to solve a problem in a faster and more efficient fashion than traditional methods by sacrificing optimality, accuracy, precision, or completeness for speed". This is in fact a more efficient solution when compared to brute force. I watched Bari's explanation regarding this algorithm on <https://www.youtube.com/watch?v=yV1d-b_NeK8>. The upper bound calculation was following what he explained in the video.

File: Algorithms/Knapsack/BranchAndBoundNode.cs
Comment (threelittle87) [reply Algorithms/Knapsack/BranchAndBoundNode.cs:64]: This algorithm was implemented in Java in my previous project. Didn't know about auto-implemented properties. Will change it. Thanks.

File: Algorithms/Knapsack/BranchAndBoundNode.cs
Comment (threelittle87) [reply Algorithms/Knapsack/BranchAndBoundNode.cs:86]: image
Initially, parent node is not declared as a nullable object, but the error(shown in picture above) shows up. The implementation requires an empty constructor in BranchAndBoundNode class, so I changed it to nullable. It's fixed right now.

File: Utilities/Exceptions/BranchAndBoundNullParentException.cs
Comment (threelittle87) [reply Utilities/Exceptions/BranchAndBoundNullParentException.cs:8]: Removed.

File: Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs
Comment (threelittle87) [reply Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs:130]: It doesn't make sense to me as well, the problem with the "non-nullable parent node must contain a non-null value when exiting a constructor" error mentioned above was not encounter when I implement this algorithm in Java. 
image
image
Looking at the first picture, although the currentNode is never null, but because is it nullable, I have to add an if statement to check if it is null. Second picture shows the original implementation, it was using a while loop, but due to the same reason I have changed it to a for loop which loops for the same amount of times and produces the same result.

File: Algorithms.Tests/Knapsack/BranchAndBoundKnapsackSolverTests.cs
Comment (threelittle87) [reply Algorithms.Tests/Knapsack/BranchAndBoundKnapsackSolverTests.cs:30]: Changed it.

**Comment by siriak** [comment]
Could you give me permission to push to your repo? I'd like to add some tests and cleanups

File: Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs
Comment (siriak) [reply Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs]: The key point is 'by sacrificing optimality'. I've found what is missing in the current implementation to give optimal solutions. This is step 1 from https://www.geeksforgeeks.org/implementation-of-0-1-knapsack-using-branch-and-bound/. I've implemented a fix and waiting on you to give me push access to your repo to push updates. (I don't want to make a fork of a fork :) to then make a PR for another PR)

File: Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs
Comment (siriak) [reply Algorithms/Knapsack/BranchAndBoundKnapsackSolver.cs:132]: I've updated this too

**Comment by threelittle87** [comment]
> Could you give me permission to push to your repo? I'd like to add some tests and cleanups

Hi, I've invited you as a collaborator for the forked repo. Thank you so much for helping me in this contribution.

**Comment by hdkfc** [comment]
i am sorry I am a fresh man. i just want to study some c# algorithms. I hope you can find another friend to help you.



---Original---
From: "Wo Zi ***@***.***&gt;
Date: Wed, Aug 4, 2021 11:56 AM
To: ***@***.***&gt;;
Cc: ***@***.***&gt;;
Subject: Re: [TheAlgorithms/C-Sharp] Branch and bound (#230)




  
Could you give me permission to push to your repo? I'd like to add some tests and cleanups
  
Hi, I've invited you as a collaborator for the forked repo. Thank you so much for helping me in this contribution.

**Comment by threelittle87** [comment]
@hdkfc Wait... What? I've sent a collab invitation to @siriak, so that he can push his changes to my repo, and I thank for his help. Did I accidentally send the invitation to you?

**Review by siriak — APPROVED** [review]
LGTM. Thanks once again for adding this algorithm, discrete optimization algorithms are probably my favorite piece of mathematics and I really enjoyed going over this one. Please review the changes I made and I'll merge if everything's ok.

**Comment by threelittle87** [comment]
I am glad that you liked this algorithm.  I think everything is ok. It's at least 10 times better than my old one. I've learnt a lot contributing this algorithm. Thanks again!

**Pull request merged**
